hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java

```diff
@@ -265,11 +270,66 @@ private void complete() {
     }
 
     /** Reruns this operation with next {@link #runStage}*/
-    private void scheduleNextRun(RunStage nextState) {
-        runStage = nextState;
+    private void scheduleNextRun(RunStage nextStage) {
+        runStage = nextStage;
+        // since this operation is being rescheduled, its invocation time and call timeout
+        // need to be recalculated (instead of allocating & scheduling a clone)
+        resetCallTimeout();
         getNodeEngine().getOperationService().execute(this);
     }
 
+    private void resetCallTimeout() {
+        long now = getNodeEngine().getClusterService().getClusterTime();
+        long memberHeartbeatTimeoutMillis = getNodeEngine().getProperties()
+                .getMillis(ClusterProperty.MAX_NO_HEARTBEAT_SECONDS);
+        long callTimeout = now + memberHeartbeatTimeoutMillis;
+        OperationAccessor.setInvocationTime(this, now);
+        OperationAccessor.setCallTimeout(this, callTimeout);
+    }
+
+    @Override
+    public void onExecutionFailure(Throwable e) {
```

Comment:
can you please remind what was the rationale of implementing `onExecutionFailure` here, since it is not called on timeout?